Support for negative prices and energy delivery

Pluimvee opened the pull request:
I've added support for negative dynamic prices and decreasing energy usage. The latter is usefull for not only net P1 readings, but also for energy readings from home-batteries.

Additionally added a recalculation on each price change event (each hour), overwriting (and correcting) small roundings issues which occur in the normal loop triggered on each energy reading

dunxd commented:
That is awesome - I'm on Octopus Agile in the UK and it has negative rates from time to time. I hadn't considered whether this was supported or not. Definitely of interest when charging my EV.

dunxd commented on custom_components/dynamic_energy_cost/energy_based_sensors.py:
How does this handle price changes more frequent than hourly? Octopus Agile tarrif in the UK changes the rate every 30 minutes and it can be quite a big jump up or down.

Pluimvee replied:
It will recalibrate on each price change event, so in your case each half hour. On more frequent price changes the recalibration benefit is no longer there en may need to be disabled. So far I noticed the recalibration benefit is into the micro cents so it may be a bit of goldplating

martinarva commented:
I think we have to solve it with a bit more complex way. Issue is that at least in our region when you buy electricity total price will be nordpool + tariff + VAT (export), but when you sell electricity its nordpoool wo. VAT (import).

Now at the moment only one price is given to the integration which in my case is total electricity price. 

Cost can be negative in two cases:
1. Because of electricity price is negative. In this case yes, when total electricity price (nordpool + tariffs + vat) is negative, this device is "earning money" and it's proper to use the total electricity price
2. When the consumption is negative. In this case consumption should be multiplied with the export price (nordpool wo. VAT) to get the right "earning". For example using grid sensor here.

So to do this right and support all scenarios this should be done:
1. User must insert both import and export price entities in configuration screen. If the import/export price is same, then user can fill the same enity in both fields
2. There should be option in configuration screen for user to select if it would be Net consumption like in HA native Utility meter. (Enable if the source is a net meter, meaning it can both increase and decrease)
3. Logic has to be rewritten to take account the two scenarios of negative price and negative consumption, so that integration calculates with the right electricity price (import or export)

It looks a bit overwhelming for me, so any contributions are welcome.

dunxd commented:
This makes sense if the integration is being used to measure the total energy spend for a household, but not my use case where I am only using it to measure the cost of running specific devices such as my EV. In those cases is the export price ever relevant for that use case?

I'm quite happy using HA's built in Energy functions for calculating my overall spend, and just use this integration to see how much certain appliances are costing to run on a highly variable tariff that can go negative. I don't export energy so I can't say whether HA is good at handling that.

Pluimvee commented:
I understand the complexity and as such I would suggest to leave all these scenarios out of this component. If prices differ per inbound and outbound I recommend users to setup two dynamic pricing components and sum these using an helper. I currently to the same as I'm comparing my energy usage with multiple energy suppliers and different contract schemes.

On the same subject: I suggest to remove the realtimesensor as this can be implemented by a simple template helper. I also suggest to remove the monthly and yearly sensors as these can be implemented using utility meters on top of this dynamic sensor component.

Apply the KISS principle

martinarva commented:
@Pluimvee, thank you very much for contributing. I’ve been so busy that I didn’t have time to maintain the integration.

I think your approach looks good, so let’s proceed with it. Maybe in the future, we can think about adding import and export prices.

Since I’m still quite busy, I’ve added access to the repository for you, so you can make the commit. It looks like there’s a conflict because of another pull request.

Also, I would be happy if you could co-manage this integration. There are some other issues that need attention as well.

fwestenberg commented:
@Pluimvee, another approach would be to keep this integration as straight forward as possible and create a helper sensor (template) to handle the specific costs. Then use this sensor for this integration… Would that be an option? What else should be added to this integration to make it work for you?

Pluimvee commented:
@fwestenberg I think you're suggesting exactly the same as I am: support negative prices and leave it up to helper templates to determine what the prices are ( with or without taxes, etc). As such the integration remains straightforward.

This is how I currently use the integration (my fork), with additional helpers to determine and compare 
- fixed price contracts
- fixed price contracts with delivery costs
- dynamic price contract with tax compensation (in NL: salderen)
- dynamic price contract without tax compensation 

In this setup I currently see the price difference per contract type, per hour, per consuming system (heat-pump, EV charger,  oven, etc)

Pluimvee commented:
I will pull the latest code and re-apply the 
1. support for negative costs 
2. the calibration logic using price events to prevent rounding issues caused by energy events (issue #37)

BDVGitHub commented:
@Pluimvee thx would be great.

Ayuus commented:
When is this update to be expected? I tried using the modified files, but appearantly, the PR is too far behind.

Pluimvee commented:
Indeed @Ayuus the PR is too far behind to be applied, or to be easily re-applied. I looked into re-applying the changes however the additional changes make it hard to pinpoint the correct locations. 

Moreover, the additional changes made do not align with keeping the component straigth forward. For example:

1. Energy prices are applicable to energy, and this should not be confused with power. Power is Watts (=Joule/sec) while energy is Joules (=kWh). If there is a need to apply this module to a Power sensor it should first be converted using an integral template helper. So in my opinion the Power sensor should be removed.

2. All the different cost trackers (daily, weekly, monthly, yearly) are created by creating additional instances of the same class. This is a nice reuse of code but all these objects create seperate event listeners with lots of CPU overhead on the load of energy and time events. I prefer to have these trackers in one object, or only have one increasing sensor which can be cleared by a configurable timer and service call.

3. Meanwhile I have also added a reconfiguration flow, an icon, and support for Joule and Cal sensors (SensorDeviceClass.ENERGY types).  

As the above include breaking changes I may better create a fork
@fwestenberg any comments?

dunxd commented:
The entire reason I started using this integration was because figuring out how to use helpers to convert power to energy was confusing. Don't throw that out. This was the only thing I found that made that stupid easy.

There are a lot of devices out there that only present power sensors.

Pluimvee commented:
> The entire reason I started using this integration was because figuring out how to use helpers to convert power to energy was confusing. Don't throw that out. This was the only thing I found that made that stupid easy.
> 
> There are a lot of devices out there that only present power sensors.

See the following on how to convert power into energy
https://www.home-assistant.io/integrations/integration

You can use the GUI to guide you or write the below in the configuration.yaml
```
sensor:
  - platform: integration
    source: sensor.current_power
    name: energy_spent
    unit_prefix: k
    round: 2

````

fwestenberg commented:
> Indeed @Ayuus the PR is too far behind to be applied, or to be easily re-applied. I looked into re-applying the changes however the last code make it hard to pinpoint the correct locations.
> 
> Moreover, the additional changes made do not align with keeping the component straigth forward. For example:
> 
> 1. Energy prices are applicable to energy, and this should not be confused with power. Power is Watts (=Joule/sec) while energy is Joules (=kWh). If there is a need to apply this module to a Power sensor it should first be converted using an integral template helper. So in my opinion the Power sensor should be removed.
> 2. All the different cost trackers (daily, weekly, monthly, yearly) are created by creating additional instances of the same class. This is a nice reuse of code but all these objects create seperate event listeners with lots of CPU overhead on the frequent energy events. I prefer to have these trackers in one object, or only have one increasing sensor which can be cleared by a configurable timer and service call (suggest the last).
> 3. Meanwhile I have also added a reconfiguration flow, an icon, and support for Joule and Cal sensors (SensorDeviceClass.ENERGY types).
> 
> As the above include breaking changes I may better create a fork @fwestenberg any comments?

1. I already asked a question some time ago why there is an power-converter integrated while this is indeed already supported by HA. But if some people are very happy with it, well…
2. Very good point. I only use this integration for my car charger, with a manual reset sensor. I disabled all other sensors, so I guess this also solved the overhead problem?
3. Very nice! This can probably be put in a separate PR?

Please let me be clear: I agree to all of the above.

dunxd commented:
I know about the Reimann integration, but I don't understand it particularly, and when I was trying to achieve calculating energy usage that integration was a major stumbling block for me.  If I had been able to figure it out, then I probably never would have sought and found this integration.

For me the most valuable feature of this integration is that it is a one stop shop for calculating energy usage no matter what the sensors present.  Taking that away for the sake of neatness or efficiency is missing the point.  More people have smart plugs and other devices that report power than know or care what a Reimann integration is.  They will find this integration first, because it is clear what it does, and most likely they will find it useful as I do.  Take that feature away, and they will move on or give up.

fwestenberg commented:
So an example of how to use the Reimann integration would solve your problem?

Or maybe even add a PR to inprove the HA docs.

Ayuus commented:
I would be very happy with a fork that can accomodate negative values for energy and/or power consumption.
I created sensors that calculate net values for both (consumtion - production), but the integration still shows positive values for hourly costs. A future improvement might be that the integration allows to enter these sensors during setup.
BTW I use Tibber (in The Netherlands), where energy price for production and consumption is the same, so for me no need to have different sensors. However, that may be different in other countries.

dunxd commented:
@fwestenberg
What is the benefit for existing and future _users_ of this integration of removing a key feature? Who is it more straight forward for?

fwestenberg commented:
I am not taking a lead in this, just giving my comments. In my opinion, it has a lot of value to keep the integration as straight forward as possible. For example, the issue with float and decimal conversion would be easier to solve with only one sensor type. And as the functionality is already in HA, this just adds extra complexity.

Pluimvee commented:
> I would be very happy with a fork that can accomodate negative values for energy and/or power consumption, as the current Implementation is IMHO useless for anyone who has solarpanels/batteries.

For now you can use my fork which supports negative prices / negative consumption (=delivery). 
https://github.com/Pluimvee/dynamic_energy_cost

From a functional perspective there are no huge differences compared to this main branch

Pluimvee commented:
Closing this one

Ayuus commented:
> > I would be very happy with a fork that can accomodate negative values for energy and/or power consumption, as the current Implementation is IMHO useless for anyone who has solarpanels/batteries.
> 
> For now you can use my fork which supports negative prices / negative consumption (=delivery). It also tracks changes in the price sensor for better accuracy https://github.com/Pluimvee/dynamic_energy_cost
> 
> From a functional perspective there are no huge differences compared to this main branch

Thanks! Installed and works! Would be great to have this in main branch, as that one has some handy entities (Manual).

Pluimvee commented:
This weekend I made the changes to the main branch, but it's still untested. Happy to create a new PR and have someone test the changes.
Again, besides the support for negative prices and energy delivery it's also more accurate as it has a second event listener on the price sensor

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Ayuus ***@***.***>
Sent: Monday, March 17, 2025 9:26:33 AM
To: martinarva/dynamic_energy_cost ***@***.***>
Cc: Pluimvee ***@***.***>; State change ***@***.***>
Subject: Re: [martinarva/dynamic_energy_cost] Support for negative prices and energy delivery (PR #31)


I would be very happy with a fork that can accomodate negative values for energy and/or power consumption, as the current Implementation is IMHO useless for anyone who has solarpanels/batteries.

For now you can use my fork which supports negative prices / negative consumption (=delivery). It also tracks changes in the price sensor for better accuracy https://github.com/Pluimvee/dynamic_energy_cost

From a functional perspective there are no huge differences compared to this main branch

Thanks! Installed and works! Would be great to have this in main branch, as that one has some handy entities (Manual).

—
Reply to this email directly, view it on GitHub<https://github.com/martinarva/dynamic_energy_cost/pull/31#issuecomment-2728563939>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/A5U6JW6WN7VQEKRLIAT4BZT2U2BLTAVCNFSM6AAAAABUSNBICGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRYGU3DGOJTHE>.
You are receiving this because you modified the open/close state.Message ID: ***@***.***>

[Ayuus]Ayuus left a comment (martinarva/dynamic_energy_cost#31)<https://github.com/martinarva/dynamic_energy_cost/pull/31#issuecomment-2728563939>

I would be very happy with a fork that can accomodate negative values for energy and/or power consumption, as the current Implementation is IMHO useless for anyone who has solarpanels/batteries.

For now you can use my fork which supports negative prices / negative consumption (=delivery). It also tracks changes in the price sensor for better accuracy https://github.com/Pluimvee/dynamic_energy_cost

From a functional perspective there are no huge differences compared to this main branch

Thanks! Installed and works! Would be great to have this in main branch, as that one has some handy entities (Manual).

—
Reply to this email directly, view it on GitHub<https://github.com/martinarva/dynamic_energy_cost/pull/31#issuecomment-2728563939>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/A5U6JW6WN7VQEKRLIAT4BZT2U2BLTAVCNFSM6AAAAABUSNBICGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRYGU3DGOJTHE>.
You are receiving this because you modified the open/close state.Message ID: ***@***.***>

Ayuus commented:
Why not bring improved accuracy to main branch, too?
I will be happy to test!

Pluimvee commented:
New PR in https://github.com/martinarva/dynamic_energy_cost/pull/79
